### PR TITLE
feat(speculative): report simulated accept length during EAGLE-3 training

### PR DESCRIPTION
**Pull request by khazic**
## What

Adds a training-time **simulated accept length** (`tau_sim`) metric to the EAGLE-3 recipe, the "Training-time acceptance-length metric for EAGLE-3" item from the tracking issue #2958. No engine, no extra forward, no new config knob: the metric is derived from counts the trainer already computes.

## How

- `Eagle3TrainerModule` now returns per-TTT-step top-1 hit / supervised-position counts (`step_correct` / `step_valid`, `[ttt_steps]` tensors) on `Eagle3StepMetrics`; the aggregate `accuracy` / `valid_tokens` are derived from their sums, so the numbers are unchanged.
- New `simulated_accept_length` helper computes the greedy chain expectation `tau = 1 + sum_k prod_{i<=k} acc_i`. Step `i`'s accuracy approximates the probability that the i-th drafted token is accepted given every earlier one was; the leading 1 is the token the target emits on every verification round, matching the `accept_length` convention of `bench_sglang` / `bench_vllm` (`1 + accepted/drafts`) and the DSpark `tau` naming in #2957.
- The recipe accumulates the counts over each logging window and sum-reduces them across ranks (counts are extensive, unlike the mean-reduced loss/accuracy), then reports `train_tau_sim` on the progress bar, the periodic and trailing-flush log lines, and W&B (`train/tau_sim`). `_run_eval` gains the same `val_tau_sim` column.
- P-EAGLE drafts all depths in parallel from one anchor rather than as a TTT chain, so the chain formula does not apply; its trainer leaves the new optional fields `None` and its logging is unchanged (covered by a test).

This is a proxy for greedy chain decoding; engine tree drafting typically accepts more. It is meant for tracking training progress in accept-length units, not for replacing the serving benchmarks.

## Tests

- `tests/unit_tests/speculative/test_eagle3.py`: hand-computed `simulated_accept_length` values (perfect draft, partial, zero-valid step truncation, hopeless draft); trainer forward exposes `[ttt_steps]` counts consistent with the aggregate accuracy/valid_tokens, monotone non-increasing valid counts.
- `tests/unit_tests/recipes/llm/test_train_eagle3_tau.py`: `_all_reduce_sum` / `_window_tau_sim` (None inputs, empty window, value, buffer non-mutation); train loop reports tau in pbar postfix and W&B for both the periodic log and the trailing flush; metrics without step counts (P-EAGLE shape) omit tau everywhere; `_run_eval` returns the tau column and `None` without step counts.

`tests/unit_tests/speculative/` + `tests/unit_tests/recipes/llm/` pass locally (the pre-existing `test_benchmark.py` failures reproduce on clean main and are unrelated). `ruff format` / `ruff check` clean.

### CODE REVIEW

**Comment by khazic** [comment]
Full log of the GPU validation run (1x A800-80GB, Qwen3-4B target, PerfectBlend train[:64] / val slice, ttt_steps=7, eager draft, micro_batch=1, grad_accum=1, log_every_steps=1, packing off, checkpointing off). Internal filesystem prefixes are redacted to $WORK; everything else is verbatim. The step lines keep the tqdm prefix, so the progress-bar tau column is visible next to the log-line train_tau_sim.

<details><summary>train.log (INFO lines, 64 steps + eval)</summary>

```
2026-07-07 22:19:43 | INFO | nemo_automodel.components.loggers.log_utils | Setting logging level to 20
2026-07-07 22:19:56 | INFO | nemo_automodel._transformers.kernel_patches | Patched model with SDPA method= [<SDPBackend.CUDNN_ATTENTION: 3>, <SDPBackend.FLASH_ATTENTION: 1>, <SDPBackend.EFFICIENT_ATTENTION: 2>, <SDPBackend.MATH: 0>]
2026-07-07 22:20:13 | INFO | root | Model summary:
2026-07-07 22:20:13 | INFO | root | --------------------------------
2026-07-07 22:20:13 | INFO | root | Trainable parameters: 4,022,468,096
2026-07-07 22:20:13 | INFO | root | Total parameters: 4,022,468,096
2026-07-07 22:20:13 | INFO | root | Trainable parameters percentage: 100.00%
2026-07-07 22:20:13 | INFO | root | Param L2 norm: 1578.0900
2026-07-07 22:20:13 | INFO | root | --------------------------------
2026-07-07 22:20:15 | INFO | nemo_automodel.components.datasets.llm.chat_dataset | Loading local dataset from $WORK/datasets/frankleeeee/PerfectBlend-Regenerated-Llama-3.1-8B-Instruct/messages via load_dataset
2026-07-07 22:20:27 | INFO | nemo_automodel.components.datasets.llm.chat_dataset | Loading local dataset from $WORK/datasets/frankleeeee/PerfectBlend-Regenerated-Llama-3.1-8B-Instruct/messages via load_dataset
2026-07-07 22:20:27 | INFO | nemo_automodel.components.datasets.llm.eagle3 | Counted 33154 supervised tokens across 4566 unique target tokens
2026-07-07 22:20:28 | INFO | nemo_automodel.components.datasets.llm.eagle3 | top 32000 token frequency ratio: 100.00%
2026-07-07 22:20:37 | INFO | root | Draft summary:
2026-07-07 22:20:37 | INFO | root | --------------------------------
2026-07-07 22:20:37 | INFO | root | Trainable parameters: 218,245,120
2026-07-07 22:20:37 | INFO | root | Total parameters: 607,201,280
2026-07-07 22:20:37 | INFO | root | Trainable parameters percentage: 35.94%
2026-07-07 22:20:37 | INFO | root | Param L2 norm: 531.6239
2026-07-07 22:20:37 | INFO | root | --------------------------------
2026-07-07 22:20:37 | INFO | __main__ | Training start: start_epoch=0 num_epochs=1 batches_per_epoch=64 grad_accum=1 log_every=1 total_optim_steps=64 warmup_steps=3 peak_lr=1.000e-04 min_lr_ratio=0.1
Training:   2%|▏         | 1/64 [00:03<03:24,  3.25s/step, acc=0.0000, loss=10.9573, lr=6.67e-05, tau=1.00]2026-07-07 22:20:40 | INFO | __main__ | epoch=0 step=1 train_loss=10.957310 train_acc=0.000000 train_tau_sim=1.0000 lr=6.667e-05
Training:   3%|▎         | 2/64 [00:03<01:31,  1.48s/step, acc=0.0000, loss=11.1715, lr=1.00e-04, tau=1.00]2026-07-07 22:20:41 | INFO | __main__ | epoch=0 step=2 train_loss=11.171521 train_acc=0.000000 train_tau_sim=1.0000 lr=1.000e-04
Training:   5%|▍         | 3/64 [00:03<00:58,  1.04step/s, acc=0.0130, loss=10.1846, lr=1.00e-04, tau=1.00]2026-07-07 22:20:41 | INFO | __main__ | epoch=0 step=3 train_loss=10.184636 train_acc=0.012987 train_tau_sim=1.0000 lr=1.000e-04
Training:   6%|▋         | 4/64 [00:04<00:42,  1.42step/s, acc=0.0242, loss=8.8750, lr=9.99e-05, tau=1.00] 2026-07-07 22:20:41 | INFO | __main__ | epoch=0 step=4 train_loss=8.874963 train_acc=0.024159 train_tau_sim=1.0021 lr=9.994e-05
Training:   8%|▊         | 5/64 [00:04<00:33,  1.76step/s, acc=0.0179, loss=8.4767, lr=9.98e-05, tau=1.04]2026-07-07 22:20:42 | INFO | __main__ | epoch=0 step=5 train_loss=8.476652 train_acc=0.017925 train_tau_sim=1.0385 lr=9.976e-05
Training:   9%|▉         | 6/64 [00:04<00:28,  2.05step/s, acc=0.0191, loss=9.8994, lr=9.95e-05, tau=1.03]2026-07-07 22:20:42 | INFO | __main__ | epoch=0 step=6 train_loss=9.899357 train_acc=0.019137 train_tau_sim=1.0252 lr=9.946e-05
Training:  11%|█         | 7/64 [00:05<00:24,  2.31step/s, acc=0.0213, loss=8.5477, lr=9.90e-05, tau=1.03]2026-07-07 22:20:42 | INFO | __main__ | epoch=0 step=7 train_loss=8.547668 train_acc=0.021277 train_tau_sim=1.0269 lr=9.905e-05
Training:  12%|█▎        | 8/64 [00:05<00:22,  2.51step/s, acc=0.0137, loss=7.9874, lr=9.85e-05, tau=1.03]2026-07-07 22:20:43 | INFO | __main__ | epoch=0 step=8 train_loss=7.987360 train_acc=0.013699 train_tau_sim=1.0347 lr=9.852e-05
Training:  14%|█▍        | 9/64 [00:05<00:20,  2.65step/s, acc=0.0305, loss=8.7068, lr=9.79e-05, tau=1.06]2026-07-07 22:20:43 | INFO | __main__ | epoch=0 step=9 train_loss=8.706793 train_acc=0.030491 train_tau_sim=1.0603 lr=9.787e-05
Training:  16%|█▌        | 10/64 [00:06<00:19,  2.77step/s, acc=0.0189, loss=9.5356, lr=9.71e-05, tau=1.04]2026-07-07 22:20:43 | INFO | __main__ | epoch=0 step=10 train_loss=9.535634 train_acc=0.018861 train_tau_sim=1.0402 lr=9.711e-05
Training:  17%|█▋        | 11/64 [00:06<00:18,  2.85step/s, acc=0.0887, loss=6.3674, lr=9.62e-05, tau=1.14]2026-07-07 22:20:43 | INFO | __main__ | epoch=0 step=11 train_loss=6.367382 train_acc=0.088670 train_tau_sim=1.1409 lr=9.623e-05
Training:  19%|█▉        | 12/64 [00:06<00:17,  2.92step/s, acc=0.0309, loss=8.1011, lr=9.53e-05, tau=1.04]2026-07-07 22:20:44 | INFO | __main__ | epoch=0 step=12 train_loss=8.101149 train_acc=0.030921 train_tau_sim=1.0361 lr=9.525e-05
Training:  20%|██        | 13/64 [00:07<00:17,  2.90step/s, acc=0.0467, loss=8.4323, lr=9.42e-05, tau=1.06]2026-07-07 22:20:44 | INFO | __main__ | epoch=0 step=13 train_loss=8.432325 train_acc=0.046732 train_tau_sim=1.0551 lr=9.416e-05
Training:  22%|██▏       | 14/64 [00:07<00:17,  2.94step/s, acc=0.0262, loss=9.1172, lr=9.30e-05, tau=1.04]2026-07-07 22:20:44 | INFO | __main__ | epoch=0 step=14 train_loss=9.117183 train_acc=0.026200 train_tau_sim=1.0385 lr=9.297e-05
Training:  23%|██▎       | 15/64 [00:07<00:16,  2.97step/s, acc=0.0865, loss=7.3575, lr=9.17e-05, tau=1.11]2026-07-07 22:20:45 | INFO | __main__ | epoch=0 step=15 train_loss=7.357465 train_acc=0.086492 train_tau_sim=1.1073 lr=9.168e-05
Training:  25%|██▌       | 16/64 [00:08<00:16,  2.99step/s, acc=0.0182, loss=9.2185, lr=9.03e-05, tau=1.01]2026-07-07 22:20:45 | INFO | __main__ | epoch=0 step=16 train_loss=9.218546 train_acc=0.018182 train_tau_sim=1.0148 lr=9.029e-05
Training:  27%|██▋       | 17/64 [00:08<00:15,  3.01step/s, acc=0.0282, loss=8.4021, lr=8.88e-05, tau=1.06]2026-07-07 22:20:45 | INFO | __main__ | epoch=0 step=17 train_loss=8.402054 train_acc=0.028222 train_tau_sim=1.0586 lr=8.880e-05
Training:  28%|██▊       | 18/64 [00:08<00:15,  3.04step/s, acc=0.0526, loss=7.6236, lr=8.72e-05, tau=1.08]2026-07-07 22:20:46 | INFO | __main__ | epoch=0 step=18 train_loss=7.623625 train_acc=0.052590 train_tau_sim=1.0825 lr=8.723e-05
Training:  30%|██▉       | 19/64 [00:09<00:14,  3.05step/s, acc=0.0701, loss=6.9957, lr=8.56e-05, tau=1.10]2026-07-07 22:20:46 | INFO | __main__ | epoch=0 step=19 train_loss=6.995727 train_acc=0.070081 train_tau_sim=1.1008 lr=8.557e-05
Training:  31%|███▏      | 20/64 [00:09<00:14,  3.04step/s, acc=0.0525, loss=8.0285, lr=8.38e-05, tau=1.08]2026-07-07 22:20:46 | INFO | __main__ | epoch=0 step=20 train_loss=8.028533 train_acc=0.052482 train_tau_sim=1.0782 lr=8.383e-05
Training:  33%|███▎      | 21/64 [00:09<00:14,  3.05step/s, acc=0.0513, loss=6.7287, lr=8.20e-05, tau=1.10]2026-07-07 22:20:47 | INFO | __main__ | epoch=0 step=21 train_loss=6.728723 train_acc=0.051318 train_tau_sim=1.1046 lr=8.201e-05
Training:  34%|███▍      | 22/64 [00:10<00:13,  3.03step/s, acc=0.0685, loss=7.9861, lr=8.01e-05, tau=1.10]2026-07-07 22:20:47 | INFO | __main__ | epoch=0 step=22 train_loss=7.986096 train_acc=0.068534 train_tau_sim=1.1012 lr=8.012e-05
Training:  36%|███▌      | 23/64 [00:10<00:13,  3.03step/s, acc=0.0647, loss=8.3794, lr=7.82e-05, tau=1.09]2026-07-07 22:20:47 | INFO | __main__ | epoch=0 step=23 train_loss=8.379358 train_acc=0.064666 train_tau_sim=1.0895 lr=7.817e-05
Training:  38%|███▊      | 24/64 [00:10<00:13,  3.05step/s, acc=0.0417, loss=8.0911, lr=7.61e-05, tau=1.09]2026-07-07 22:20:48 | INFO | __main__ | epoch=0 step=24 train_loss=8.091060 train_acc=0.041667 train_tau_sim=1.0902 lr=7.615e-05
Training:  39%|███▉      | 25/64 [00:11<00:12,  3.06step/s, acc=0.0649, loss=6.5258, lr=7.41e-05, tau=1.19]2026-07-07 22:20:48 | INFO | __main__ | epoch=0 step=25 train_loss=6.525767 train_acc=0.064855 train_tau_sim=1.1910 lr=7.408e-05
Training:  41%|████      | 26/64 [00:11<00:12,  3.07step/s, acc=0.0569, loss=6.6924, lr=7.20e-05, tau=1.15]2026-07-07 22:20:48 | INFO | __main__ | epoch=0 step=26 train_loss=6.692424 train_acc=0.056928 train_tau_sim=1.1461 lr=7.195e-05
Training:  42%|████▏     | 27/64 [00:11<00:12,  3.08step/s, acc=0.0921, loss=6.2703, lr=6.98e-05, tau=1.21]2026-07-07 22:20:49 | INFO | __main__ | epoch=0 step=27 train_loss=6.270345 train_acc=0.092083 train_tau_sim=1.2123 lr=6.978e-05
Training:  44%|████▍     | 28/64 [00:12<00:11,  3.08step/s, acc=0.0491, loss=8.0298, lr=6.76e-05, tau=1.07]2026-07-07 22:20:49 | INFO | __main__ | epoch=0 step=28 train_loss=8.029823 train_acc=0.049051 train_tau_sim=1.0712 lr=6.758e-05
Training:  45%|████▌     | 29/64 [00:12<00:11,  3.08step/s, acc=0.0926, loss=5.5240, lr=6.53e-05, tau=1.20]2026-07-07 22:20:49 | INFO | __main__ | epoch=0 step=29 train_loss=5.523967 train_acc=0.092622 train_tau_sim=1.2032 lr=6.534e-05
Training:  47%|████▋     | 30/64 [00:12<00:11,  3.08step/s, acc=0.0747, loss=7.4043, lr=6.31e-05, tau=1.15]2026-07-07 22:20:50 | INFO | __main__ | epoch=0 step=30 train_loss=7.404286 train_acc=0.074695 train_tau_sim=1.1511 lr=6.307e-05
Training:  48%|████▊     | 31/64 [00:13<00:10,  3.08step/s, acc=0.0697, loss=7.2529, lr=6.08e-05, tau=1.14]2026-07-07 22:20:50 | INFO | __main__ | epoch=0 step=31 train_loss=7.252875 train_acc=0.069698 train_tau_sim=1.1377 lr=6.078e-05
Training:  50%|█████     | 32/64 [00:13<00:10,  3.06step/s, acc=0.0581, loss=7.7637, lr=5.85e-05, tau=1.10]2026-07-07 22:20:50 | INFO | __main__ | epoch=0 step=32 train_loss=7.763668 train_acc=0.058085 train_tau_sim=1.0974 lr=5.847e-05
Training:  52%|█████▏    | 33/64 [00:13<00:10,  3.06step/s, acc=0.0555, loss=7.2672, lr=5.62e-05, tau=1.11]2026-07-07 22:20:51 | INFO | __main__ | epoch=0 step=33 train_loss=7.267240 train_acc=0.055540 train_tau_sim=1.1144 lr=5.616e-05
Training:  53%|█████▎    | 34/64 [00:14<00:09,  3.05step/s, acc=0.1132, loss=5.9095, lr=5.38e-05, tau=1.19]2026-07-07 22:20:51 | INFO | __main__ | epoch=0 step=34 train_loss=5.909540 train_acc=0.113208 train_tau_sim=1.1869 lr=5.384e-05
Training:  55%|█████▍    | 35/64 [00:14<00:09,  3.05step/s, acc=0.0622, loss=7.1938, lr=5.15e-05, tau=1.09]2026-07-07 22:20:51 | INFO | __main__ | epoch=0 step=35 train_loss=7.193817 train_acc=0.062192 train_tau_sim=1.0872 lr=5.153e-05
Training:  56%|█████▋    | 36/64 [00:14<00:09,  3.05step/s, acc=0.0429, loss=8.3937, lr=4.92e-05, tau=1.08]2026-07-07 22:20:52 | INFO | __main__ | epoch=0 step=36 train_loss=8.393723 train_acc=0.042877 train_tau_sim=1.0803 lr=4.922e-05
Training:  58%|█████▊    | 37/64 [00:15<00:08,  3.06step/s, acc=0.0673, loss=7.2704, lr=4.69e-05, tau=1.12]2026-07-07 22:20:52 | INFO | __main__ | epoch=0 step=37 train_loss=7.270431 train_acc=0.067323 train_tau_sim=1.1166 lr=4.693e-05
Training:  59%|█████▉    | 38/64 [00:15<00:08,  3.07step/s, acc=0.0107, loss=8.2841, lr=4.47e-05, tau=1.03]2026-07-07 22:20:52 | INFO | __main__ | epoch=0 step=38 train_loss=8.284101 train_acc=0.010703 train_tau_sim=1.0328 lr=4.466e-05
Training:  61%|██████    | 39/64 [00:15<00:08,  3.07step/s, acc=0.1136, loss=5.5930, lr=4.24e-05, tau=1.18]2026-07-07 22:20:53 | INFO | __main__ | epoch=0 step=39 train_loss=5.593010 train_acc=0.113579 train_tau_sim=1.1837 lr=4.242e-05
Training:  62%|██████▎   | 40/64 [00:16<00:07,  3.06step/s, acc=0.0597, loss=7.9108, lr=4.02e-05, tau=1.09]2026-07-07 22:20:53 | INFO | __main__ | epoch=0 step=40 train_loss=7.910803 train_acc=0.059719 train_tau_sim=1.0940 lr=4.022e-05
Training:  64%|██████▍   | 41/64 [00:16<00:07,  3.01step/s, acc=0.0666, loss=7.7597, lr=3.80e-05, tau=1.08]2026-07-07 22:20:53 | INFO | __main__ | epoch=0 step=41 train_loss=7.759747 train_acc=0.066619 train_tau_sim=1.0773 lr=3.805e-05
Training:  66%|██████▌   | 42/64 [00:16<00:07,  3.02step/s, acc=0.1011, loss=6.3140, lr=3.59e-05, tau=1.16]2026-07-07 22:20:54 | INFO | __main__ | epoch=0 step=42 train_loss=6.313971 train_acc=0.101078 train_tau_sim=1.1591 lr=3.592e-05
Training:  67%|██████▋   | 43/64 [00:17<00:06,  3.04step/s, acc=0.0860, loss=5.7772, lr=3.39e-05, tau=1.17]2026-07-07 22:20:54 | INFO | __main__ | epoch=0 step=43 train_loss=5.777194 train_acc=0.085973 train_tau_sim=1.1724 lr=3.385e-05
Training:  69%|██████▉   | 44/64 [00:17<00:06,  3.05step/s, acc=0.1027, loss=5.8126, lr=3.18e-05, tau=1.24]2026-07-07 22:20:54 | INFO | __main__ | epoch=0 step=44 train_loss=5.812594 train_acc=0.102729 train_tau_sim=1.2449 lr=3.183e-05
Training:  70%|███████   | 45/64 [00:17<00:06,  3.01step/s, acc=0.0767, loss=7.4826, lr=2.99e-05, tau=1.12]2026-07-07 22:20:55 | INFO | __main__ | epoch=0 step=45 train_loss=7.482585 train_acc=0.076748 train_tau_sim=1.1207 lr=2.988e-05
Training:  72%|███████▏  | 46/64 [00:18<00:05,  3.03step/s, acc=0.0939, loss=6.1569, lr=2.80e-05, tau=1.21]2026-07-07 22:20:55 | INFO | __main__ | epoch=0 step=46 train_loss=6.156881 train_acc=0.093878 train_tau_sim=1.2111 lr=2.799e-05
Training:  73%|███████▎  | 47/64 [00:18<00:05,  3.04step/s, acc=0.0938, loss=6.3294, lr=2.62e-05, tau=1.17]2026-07-07 22:20:55 | INFO | __main__ | epoch=0 step=47 train_loss=6.329416 train_acc=0.093750 train_tau_sim=1.1727 lr=2.617e-05
Training:  75%|███████▌  | 48/64 [00:18<00:05,  3.06step/s, acc=0.0695, loss=6.6279, lr=2.44e-05, tau=1.15]2026-07-07 22:20:56 | INFO | __main__ | epoch=0 step=48 train_loss=6.627916 train_acc=0.069472 train_tau_sim=1.1522 lr=2.443e-05
Training:  77%|███████▋  | 49/64 [00:18<00:04,  3.06step/s, acc=0.0430, loss=7.7133, lr=2.28e-05, tau=1.09]2026-07-07 22:20:56 | INFO | __main__ | epoch=0 step=49 train_loss=7.713265 train_acc=0.043033 train_tau_sim=1.0903 lr=2.277e-05
Training:  78%|███████▊  | 50/64 [00:19<00:04,  2.99step/s, acc=0.0332, loss=7.6939, lr=2.12e-05, tau=1.07]2026-07-07 22:20:56 | INFO | __main__ | epoch=0 step=50 train_loss=7.693886 train_acc=0.033172 train_tau_sim=1.0659 lr=2.120e-05
Training:  80%|███████▉  | 51/64 [00:19<00:04,  3.01step/s, acc=0.1026, loss=5.5871, lr=1.97e-05, tau=1.29]2026-07-07 22:20:57 | INFO | __main__ | epoch=0 step=51 train_loss=5.587146 train_acc=0.102616 train_tau_sim=1.2948 lr=1.971e-05
Training:  81%|████████▏ | 52/64 [00:19<00:03,  3.04step/s, acc=0.0796, loss=5.6938, lr=1.83e-05, tau=1.16]2026-07-07 22:20:57 | INFO | __main__ | epoch=0 step=52 train_loss=5.693798 train_acc=0.079566 train_tau_sim=1.1604 lr=1.832e-05
Training:  83%|████████▎ | 53/64 [00:20<00:03,  3.04step/s, acc=0.0426, loss=7.5771, lr=1.70e-05, tau=1.08]2026-07-07 22:20:57 | INFO | __main__ | epoch=0 step=53 train_loss=7.577057 train_acc=0.042644 train_tau_sim=1.0794 lr=1.703e-05
Training:  84%|████████▍ | 54/64 [00:20<00:03,  3.05step/s, acc=0.1227, loss=5.2783, lr=1.58e-05, tau=1.27]2026-07-07 22:20:58 | INFO | __main__ | epoch=0 step=54 train_loss=5.278314 train_acc=0.122723 train_tau_sim=1.2707 lr=1.584e-05
Training:  86%|████████▌ | 55/64 [00:20<00:02,  3.06step/s, acc=0.0890, loss=6.7710, lr=1.47e-05, tau=1.14]2026-07-07 22:20:58 | INFO | __main__ | epoch=0 step=55 train_loss=6.771029 train_acc=0.088993 train_tau_sim=1.1407 lr=1.475e-05
Training:  88%|████████▊ | 56/64 [00:21<00:02,  3.07step/s, acc=0.0321, loss=7.1991, lr=1.38e-05, tau=1.07]2026-07-07 22:20:58 | INFO | __main__ | epoch=0 step=56 train_loss=7.199089 train_acc=0.032060 train_tau_sim=1.0686 lr=1.377e-05
Training:  89%|████████▉ | 57/64 [00:21<00:02,  3.01step/s, acc=0.0536, loss=7.5375, lr=1.29e-05, tau=1.09]2026-07-07 22:20:59 | INFO | __main__ | epoch=0 step=57 train_loss=7.537496 train_acc=0.053634 train_tau_sim=1.0874 lr=1.289e-05
Training:  91%|█████████ | 58/64 [00:21<00:01,  3.03step/s, acc=0.0447, loss=6.3551, lr=1.21e-05, tau=1.10]2026-07-07 22:20:59 | INFO | __main__ | epoch=0 step=58 train_loss=6.355084 train_acc=0.044716 train_tau_sim=1.0958 lr=1.213e-05
Training:  92%|█████████▏| 59/64 [00:22<00:01,  3.04step/s, acc=0.0962, loss=5.8761, lr=1.15e-05, tau=1.26]2026-07-07 22:20:59 | INFO | __main__ | epoch=0 step=59 train_loss=5.876127 train_acc=0.096154 train_tau_sim=1.2638 lr=1.148e-05
Training:  94%|█████████▍| 60/64 [00:22<00:01,  3.04step/s, acc=0.0581, loss=7.7989, lr=1.10e-05, tau=1.11]2026-07-07 22:21:00 | INFO | __main__ | epoch=0 step=60 train_loss=7.798913 train_acc=0.058108 train_tau_sim=1.1141 lr=1.095e-05
Training:  95%|█████████▌| 61/64 [00:22<00:00,  3.05step/s, acc=0.0952, loss=6.1116, lr=1.05e-05, tau=1.20]2026-07-07 22:21:00 | INFO | __main__ | epoch=0 step=61 train_loss=6.111648 train_acc=0.095238 train_tau_sim=1.2010 lr=1.054e-05
Training:  97%|█████████▋| 62/64 [00:23<00:00,  3.07step/s, acc=0.1345, loss=4.5765, lr=1.02e-05, tau=1.28]2026-07-07 22:21:00 | INFO | __main__ | epoch=0 step=62 train_loss=4.576526 train_acc=0.134535 train_tau_sim=1.2765 lr=1.024e-05
Training:  98%|█████████▊| 63/64 [00:23<00:00,  3.07step/s, acc=0.0676, loss=6.3037, lr=1.01e-05, tau=1.13]2026-07-07 22:21:01 | INFO | __main__ | epoch=0 step=63 train_loss=6.303699 train_acc=0.067553 train_tau_sim=1.1312 lr=1.006e-05
Training: 100%|██████████| 64/64 [00:23<00:00,  3.08step/s, acc=0.0756, loss=7.4694, lr=1.00e-05, tau=1.11]2026-07-07 22:21:01 | INFO | __main__ | epoch=0 step=64 train_loss=7.469352 train_acc=0.075630 train_tau_sim=1.1128 lr=1.000e-05
2026-07-07 22:21:01 | INFO | __main__ | Epoch 0 done: total_batches_seen=64 global_step=64
2026-07-07 22:21:04 | INFO | __main__ | epoch=0 val_loss=6.469628 val_acc=0.079784 val_tau_sim=1.1301
2026-07-07 22:21:04 | INFO | __main__ | Training complete: global_step=64
EXIT_CODE=0
```

</details>

File: nemo_automodel/components/speculative/eagle/core.py [hunk 1]
```diff
@@ -193,14 +220,23 @@ def forward(
 
             valid_mask = step_position_mask.squeeze(-1).bool()
             correct = (logits.argmax(dim=-1) == cur_target_probs.argmax(dim=-1)) & valid_mask
-            running_correct = running_correct + correct.sum()
-            running_valid = running_valid + valid_mask.sum()
+            step_correct_parts.append(correct.sum())
```
Comment (HuiyingLi) [inline nemo_automodel/components/speculative/eagle/core.py]: These per-step reductions discard the correlation needed for greedy prefix acceptance. `simulated_accept_length` later computes `1 + E[C1] + E[C1]E[C2] + ...`, while the actual greedy expectation is `1 + E[C1] + E[C1*C2] + ...`. For example, with both depths 50% accurate, coincident hits give an actual tau of 2.0 and disjoint hits give 1.5, but the current calculation reports 1.75 for both. Since `correct` is still per-position here, could we accumulate a running prefix-correct mask and expose prefix-hit counts instead? That would preserve the no-extra-forward property.

File: nemo_automodel/components/speculative/eagle/core.py [hunk 2]
```diff
@@ -193,14 +220,23 @@ def forward(
...
+            step_valid_parts.append(valid_mask.sum())
```
Comment (HuiyingLi) [inline nemo_automodel/components/speculative/eagle/core.py]: `valid_mask` comes from `position_mask`, which also requires `selected_token_mask[target_top_ids]`. This drops target tokens outside the compressed draft vocabulary from the denominator, although serving must reject those tokens because the draft cannot emit them. That biases `tau_sim` upward whenever vocabulary coverage is below 100%. Could the acceptance denominator use the shifted loss/document-validity mask, with out-of-draft-vocabulary targets counted as zero-correct, while retaining `position_mask` for the CE loss?

**Review by HuiyingLi — COMMENTED** [review]
Two metric-correctness concerns are noted inline.

File: nemo_automodel/components/speculative/eagle/core.py (hunk 1 above)
Comment (khazic) [reply nemo_automodel/components/speculative/eagle/core.py]: Good catch, fixed in 540385f50: the trainer now AND-accumulates a per-slot prefix-correct mask across TTT steps (the shift keeps slot j anchored to the same chain) and exposes prefix-hit counts, so tau is exactly 1 + sum_k P(chain survives depth k+1). Your coincident vs disjoint example is now a unit test.

File: nemo_automodel/components/speculative/eagle/core.py (hunk 2 above)
Comment (khazic) [reply nemo_automodel/components/speculative/eagle/core.py]: Fixed in 540385f50: the acceptance denominator now uses the shifted loss/document mask, so out-of-draft-vocab targets stay in the denominator and count as rejections (correct is already coverage-masked, so they can never be hits). position_mask still gates the CE loss unchanged. Added a regression test with partial vocabulary coverage.

File: nemo_automodel/components/speculative/eagle/core.py [hunk 3]
```diff
@@ -196,11 +243,22 @@ def forward(
             running_correct = running_correct + correct.sum()
             running_valid = running_valid + valid_mask.sum()
 
+            prefix_correct = correct if prefix_correct is None else prefix_correct & correct
+            step_prefix_parts.append(prefix_correct.sum())
+            step_valid_parts.append(chain_valid.sum())
```
Comment (HuiyingLi) [inline nemo_automodel/components/speculative/eagle/core.py]: The numerator and denominator here describe different chain populations: `prefix_correct` requires every earlier shifted loss/document mask to have been valid, while `step_valid` counts only the current `chain_valid`. For a normal prompt/assistant mask `[0, 1, 1, 1]` with perfect predictions, the reported survival rates become `[1, 2/3, 1/2]`, so `tau_sim=3.1667` instead of `4.0`. This also affects multi-turn data, where `ChatDataset` explicitly permits gappy loss masks. Could we accumulate a `prefix_valid` mask in parallel with `prefix_correct` (`prefix_valid = chain_valid if first else prefix_valid & chain_valid`) and append `prefix_valid.sum()` here? A regression test with a prompt-to-assistant or gappy mask would cover the boundary case that the current all-ones tests miss.

**Review by HuiyingLi — COMMENTED** [review]
One follow-up correctness concern is noted inline.

File: nemo_automodel/components/speculative/eagle/core.py (hunk 3 above)
Comment (khazic) [reply nemo_automodel/components/speculative/eagle/core.py]: You are right, the populations diverged for non-nested masks. Fixed in a8f82aa98: a prefix_valid mask is AND-accumulated alongside prefix_correct and its count is the denominator, so chains with an unsupervised earlier depth leave both sides while out-of-vocab rejections still count as misses. Your prompt-mask example now reports tau=4.0 for a perfect draft, and a gappy-mask regression test asserts the AND-accumulated counts.

**Review by HuiyingLi — APPROVED** [review]
